File: apps/frontend/src/api/findings.ts
/**
 * Findings API Client
 */

import { useAuthStore } from '@/stores/auth'

const API_BASE = '/api'

// --- Types ---

export type FindingType = 'irregularity' | 'deficiency' | 'recommendation' | 'observation'
export type FindingStatus = 'draft' | 'confirmed' | 'disputed' | 'resolved' | 'withdrawn'
export type ErrorCategory =
  | 'ineligible_expenditure'
  | 'public_procurement'
  | 'missing_documents'
  | 'calculation_error'
  | 'double_funding'
  | 'other'

export interface Finding {
  id: string
  audit_case_id: string
  finding_number: number
  finding_type: FindingType
  error_category: ErrorCategory | null
  title: string
  description: string
  financial_impact: number | null
  is_systemic: boolean
  status: FindingStatus
  response_requested: boolean
  response_deadline: string | null
  response_received: string | null
  response_received_at: string | null
  final_assessment: string | null
  corrective_action: string | null
  created_at: string
  updated_at: string
}

export interface FindingCreate {
  finding_type: FindingType
  error_category?: ErrorCategory | null
  title: string
  description: string
  financial_impact?: number | null
  is_systemic?: boolean
  response_requested?: boolean
  response_deadline?: string | null
}

export interface FindingUpdate {
  finding_type?: FindingType
  error_category?: ErrorCategory | null
  title?: string
  description?: string
  financial_impact?: number | null
  is_systemic?: boolean
  status?: FindingStatus
  response_requested?: boolean
  response_deadline?: string | null
  response_received?: string | null
  final_assessment?: string | null
  corrective_action?: string | null
}

export interface FindingsSummary {
  total: number
  by_status: Record<string, number>
  by_type: Record<string, number>
  total_financial_impact: number
}

// --- Helpers ---

function getAuthHeaders(): Record<string, string> {
  const authStore = useAuthStore()
  return {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${authStore.token}`,
  }
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Request failed' }))
    throw new Error(error.detail || `HTTP error ${response.status}`)
  }
  return response.json()
}

// --- API Functions ---

/**
 * List findings for an audit case
 */
export async function listFindings(
  caseId: string,
  params: { status?: FindingStatus; finding_type?: FindingType } = {}
): Promise<Finding[]> {
  const queryParams = new URLSearchParams()

  if (params.status) queryParams.set('status', params.status)
  if (params.finding_type) queryParams.set('finding_type', params.finding_type)

  const url = `${API_BASE}/audit-cases/${caseId}/findings${
    queryParams.toString() ? '?' + queryParams.toString() : ''
  }`

  const response = await fetch(url, {
    headers: getAuthHeaders(),
  })

  return handleResponse<Finding[]>(response)
}

/**
 * Create a new finding
 */
export async function createFinding(caseId: string, data: FindingCreate): Promise<Finding> {
  const response = await fetch(`${API_BASE}/audit-cases/${caseId}/findings`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify(data),
  })

  return handleResponse<Finding>(response)
}

/**
 * Get a specific finding
 */
export async function getFinding(caseId: string, findingId: string): Promise<Finding> {
  const response = await fetch(`${API_BASE}/audit-cases/${caseId}/findings/${findingId}`, {
    headers: getAuthHeaders(),
  })

  return handleResponse<Finding>(response)
}

/**
 * Update a finding
 */
export async function updateFinding(
  caseId: string,
  findingId: string,
  data: FindingUpdate
): Promise<Finding> {
  const response = await fetch(`${API_BASE}/audit-cases/${caseId}/findings/${findingId}`, {
    method: 'PATCH',
    headers: getAuthHeaders(),
    body: JSON.stringify(data),
  })

  return handleResponse<Finding>(response)
}

/**
 * Delete a finding
 */
export async function deleteFinding(caseId: string, findingId: string): Promise<void> {
  const response = await fetch(`${API_BASE}/audit-cases/${caseId}/findings/${findingId}`, {
    method: 'DELETE',
    headers: getAuthHeaders(),
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to delete' }))
    throw new Error(error.detail)
  }
}

/**
 * Confirm a draft finding
 */
export async function confirmFinding(caseId: string, findingId: string): Promise<Finding> {
  const response = await fetch(`${API_BASE}/audit-cases/${caseId}/findings/${findingId}/confirm`, {
    method: 'POST',
    headers: getAuthHeaders(),
  })

  return handleResponse<Finding>(response)
}

/**
 * Mark a finding as resolved
 */
export async function resolveFinding(
  caseId: string,
  findingId: string,
  correctiveAction?: string
): Promise<Finding> {
  const queryParams = correctiveAction ? `?corrective_action=${encodeURIComponent(correctiveAction)}` : ''
  const response = await fetch(
    `${API_BASE}/audit-cases/${caseId}/findings/${findingId}/resolve${queryParams}`,
    {
      method: 'POST',
      headers: getAuthHeaders(),
    }
  )

  return handleResponse<Finding>(response)
}

/**
 * Get findings summary for an audit case
 */
export async function getFindingsSummary(caseId: string): Promise<FindingsSummary> {
  const response = await fetch(`${API_BASE}/audit-cases/${caseId}/findings/stats/summary`, {
    headers: getAuthHeaders(),
  })

  return handleResponse<FindingsSummary>(response)
}

// --- Labels ---

export const FINDING_TYPE_LABELS: Record<FindingType, string> = {
  irregularity: 'Unregelmäßigkeit',
  deficiency: 'Mangel',
  recommendation: 'Empfehlung',
  observation: 'Beobachtung',
}

export const FINDING_STATUS_LABELS: Record<FindingStatus, string> = {
  draft: 'Entwurf',
  confirmed: 'Bestätigt',
  disputed: 'Widerspruch',
  resolved: 'Behoben',
  withdrawn: 'Zurückgezogen',
}

export const ERROR_CATEGORY_LABELS: Record<ErrorCategory, string> = {
  ineligible_expenditure: 'Nicht förderfähige Ausgaben',
  public_procurement: 'Vergabefehler',
  missing_documents: 'Fehlende Unterlagen',
  calculation_error: 'Rechenfehler',
  double_funding: 'Doppelfinanzierung',
  other: 'Sonstiges',
}

export const FINDING_TYPE_COLORS: Record<FindingType, string> = {
  irregularity: 'bg-red-100 text-red-800 dark:bg-red-900 dark:text-red-300',
  deficiency: 'bg-orange-100 text-orange-800 dark:bg-orange-900 dark:text-orange-300',
  recommendation: 'bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-300',
  observation: 'bg-gray-100 text-gray-800 dark:bg-gray-800 dark:text-gray-300',
}

export const FINDING_STATUS_COLORS: Record<FindingStatus, string> = {
  draft: 'bg-gray-100 text-gray-800 dark:bg-gray-800 dark:text-gray-300',
  confirmed: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900 dark:text-yellow-300',
  disputed: 'bg-purple-100 text-purple-800 dark:bg-purple-900 dark:text-purple-300',
  resolved: 'bg-green-100 text-green-800 dark:bg-green-900 dark:text-green-300',
  withdrawn: 'bg-gray-100 text-gray-500 dark:bg-gray-800 dark:text-gray-500',
}

/**
 * Format currency amount for display
 */
export function formatCurrency(amount: number | null): string {
  if (amount === null || amount === undefined) return '-'
  return new Intl.NumberFormat('de-DE', {
    style: 'currency',
    currency: 'EUR',
  }).format(amount)
}

File: apps/backend/app/api/findings.py
"""Findings API Endpoints."""

from datetime import datetime, timezone
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.api.auth import get_current_user
from app.api.audit_logs import log_audit_event
from app.models.user import User
from app.models.audit_case import AuditCase, AuditCaseFinding
from app.schemas.audit_case import (
    FindingCreate,
    FindingUpdate,
    FindingResponse,
)

# Finding type labels for logging
FINDING_TYPE_LABELS = {
    "irregularity": "Unregelmäßigkeit",
    "deficiency": "Mangel",
    "recommendation": "Empfehlung",
    "observation": "Beobachtung",
}

router = APIRouter(prefix="/audit-cases/{case_id}/findings", tags=["Findings"])


# --- Helper Functions ---


async def get_audit_case_or_404(
    case_id: str,
    db: AsyncSession,
    user: User,
) -> AuditCase:
    """Get audit case or raise 404."""
    result = await db.execute(
        select(AuditCase).where(
            AuditCase.id == case_id,
            AuditCase.tenant_id == user.tenant_id,
        )
    )
    case = result.scalar_one_or_none()
    if not case:
        raise HTTPException(status_code=404, detail="Audit case not found")
    return case


async def get_finding_or_404(
    finding_id: str,
    case_id: str,
    db: AsyncSession,
) -> AuditCaseFinding:
    """Get finding or raise 404."""
    result = await db.execute(
        select(AuditCaseFinding).where(
            AuditCaseFinding.id == finding_id,
            AuditCaseFinding.audit_case_id == case_id,
        )
    )
    finding = result.scalar_one_or_none()
    if not finding:
        raise HTTPException(status_code=404, detail="Finding not found")
    return finding


async def get_next_finding_number(case_id: str, db: AsyncSession) -> int:
    """Get next finding number for a case."""
    result = await db.execute(
        select(func.max(AuditCaseFinding.finding_number)).where(
            AuditCaseFinding.audit_case_id == case_id
        )
    )
    max_number = result.scalar()
    return (max_number or 0) + 1


# --- Endpoints ---


@router.get("")
async def list_findings(
    case_id: str,
    status: str | None = None,
    finding_type: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[FindingResponse]:
    """List findings for an audit case."""
    await get_audit_case_or_404(case_id, db, current_user)

    query = select(AuditCaseFinding).where(AuditCaseFinding.audit_case_id == case_id)

    if status:
        query = query.where(AuditCaseFinding.status == status)
    if finding_type:
        query = query.where(AuditCaseFinding.finding_type == finding_type)

    query = query.order_by(AuditCaseFinding.finding_number)

    result = await db.execute(query)
    findings = result.scalars().all()

    return [FindingResponse.model_validate(f) for f in findings]


@router.post("", status_code=201)
async def create_finding(
    case_id: str,
    data: FindingCreate,
    request: Request,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> FindingResponse:
    """Create a new finding for an audit case."""
    case = await get_audit_case_or_404(case_id, db, current_user)

    # Get next finding number
    finding_number = await get_next_finding_number(case_id, db)

    finding = AuditCaseFinding(
        id=str(uuid4()),
        audit_case_id=case.id,
        finding_number=finding_number,
        finding_type=data.finding_type,
        error_category=data.error_category,
        title=data.title,
        description=data.description,
        financial_impact=data.financial_impact,
        is_systemic=data.is_systemic,
        status="draft",
        response_requested=data.response_requested,
        response_deadline=data.response_deadline,
    )

    db.add(finding)
    await db.flush()

    # Log the creation
    type_label = FINDING_TYPE_LABELS.get(data.finding_type, data.finding_type)
    await log_audit_event(
        db=db,
        tenant_id=current_user.tenant_id,
        entity_type="audit_case",
        entity_id=case_id,
        action="create",
        user=current_user,
        description=f"Feststellung #{finding_number} erstellt: {type_label} - {data.title}",
        request=request,
    )

    await db.commit()
    await db.refresh(finding)

    return FindingResponse.model_validate(finding)


@router.get("/{finding_id}")
async def get_finding(
    case_id: str,
    finding_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> FindingResponse:
    """Get a specific finding."""
    await get_audit_case_or_404(case_id, db, current_user)
    finding = await get_finding_or_404(finding_id, case_id, db)
    return FindingResponse.model_validate(finding)


@router.patch("/{finding_id}")
async def update_finding(
    case_id: str,
    finding_id: str,
    data: FindingUpdate,
    request: Request,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> FindingResponse:
    """Update a finding."""
    await get_audit_case_or_404(case_id, db, current_user)
    finding = await get_finding_or_404(finding_id, case_id, db)

    update_data = data.model_dump(exclude_unset=True)

    # Handle response received timestamp
    if "response_received" in update_data and update_data["response_received"]:
        update_data["response_received_at"] = datetime.now(timezone.utc)

    # Track changes for logging
    changes = {}
    for key, value in update_data.items():
        old_value = getattr(finding, key)
        if old_value != value:
            changes[key] = {
                "old": str(old_value) if old_value else None,
                "new": str(value) if value else None,
            }
        setattr(finding, key, value)

    if changes:
        await log_audit_event(
            db=db,
            tenant_id=current_user.tenant_id,
            entity_type="audit_case",
            entity_id=case_id,
            action="update",
            user=current_user,
            changes=changes,
            description=f"Feststellung #{finding.finding_number} aktualisiert",
            request=request,
        )

    await db.commit()
    await db.refresh(finding)

    return FindingResponse.model_validate(finding)


@router.delete("/{finding_id}", status_code=204)
async def delete_finding(
    case_id: str,
    finding_id: str,
    request: Request,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> None:
    """Delete a finding."""
    await get_audit_case_or_404(case_id, db, current_user)
    finding = await get_finding_or_404(finding_id, case_id, db)

    # Only allow deleting draft findings
    if finding.status != "draft":
        raise HTTPException(
            status_code=400,
            detail="Only draft findings can be deleted",
        )

    finding_number = finding.finding_number
    finding_title = finding.title

    # Log the deletion
    await log_audit_event(
        db=db,
        tenant_id=current_user.tenant_id,
        entity_type="audit_case",
        entity_id=case_id,
        action="delete",
        user=current_user,
        description=f"Feststellung #{finding_number} gelöscht: {finding_title}",
        request=request,
    )

    await db.delete(finding)
    await db.commit()


# --- Bulk Operations ---


@router.post("/{finding_id}/confirm")
async def confirm_finding(
    case_id: str,
    finding_id: str,
    request: Request,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> FindingResponse:
    """Confirm a draft finding."""
    await get_audit_case_or_404(case_id, db, current_user)
    finding = await get_finding_or_404(finding_id, case_id, db)

    if finding.status != "draft":
        raise HTTPException(
            status_code=400,
            detail="Only draft findings can be confirmed",
        )

    finding.status = "confirmed"

    # Log the confirmation
    await log_audit_event(
        db=db,
        tenant_id=current_user.tenant_id,
        entity_type="audit_case",
        entity_id=case_id,
        action="confirm",
        user=current_user,
        field_name="status",
        old_value="draft",
        new_value="confirmed",
        description=f"Feststellung #{finding.finding_number} bestätigt: {finding.title}",
        request=request,
    )

    await db.commit()
    await db.refresh(finding)

    return FindingResponse.model_validate(finding)


@router.post("/{finding_id}/resolve")
async def resolve_finding(
    case_id: str,
    finding_id: str,
    request: Request,
    corrective_action: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> FindingResponse:
    """Mark a finding as resolved."""
    await get_audit_case_or_404(case_id, db, current_user)
    finding = await get_finding_or_404(finding_id, case_id, db)

    old_status = finding.status
    if old_status not in ["confirmed", "disputed"]:
        raise HTTPException(
            status_code=400,
            detail="Only confirmed or disputed findings can be resolved",
        )

    finding.status = "resolved"
    if corrective_action:
        finding.corrective_action = corrective_action

    # Log the resolution
    await log_audit_event(
        db=db,
        tenant_id=current_user.tenant_id,
        entity_type="audit_case",
        entity_id=case_id,
        action="resolve",
        user=current_user,
        field_name="status",
        old_value=old_status,
        new_value="resolved",
        description=f"Feststellung #{finding.finding_number} behoben: {finding.title}",
        request=request,
    )

    await db.commit()
    await db.refresh(finding)

    return FindingResponse.model_validate(finding)


# --- Statistics ---


@router.get("/stats/summary")
async def get_findings_summary(
    case_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """Get findings summary for an audit case."""
    await get_audit_case_or_404(case_id, db, current_user)

    # Total count
    total_result = await db.execute(
        select(func.count()).where(AuditCaseFinding.audit_case_id == case_id)
    )
    total = total_result.scalar() or 0

    # Count by status
    status_result = await db.execute(
        select(AuditCaseFinding.status, func.count())
        .where(AuditCaseFinding.audit_case_id == case_id)
        .group_by(AuditCaseFinding.status)
    )
    by_status = dict(status_result.all())

    # Count by type
    type_result = await db.execute(
        select(AuditCaseFinding.finding_type, func.count())
        .where(AuditCaseFinding.audit_case_id == case_id)
        .group_by(AuditCaseFinding.finding_type)
    )
    by_type = dict(type_result.all())

    # Total financial impact
    impact_result = await db.execute(
        select(func.sum(AuditCaseFinding.financial_impact)).where(
            AuditCaseFinding.audit_case_id == case_id
        )
    )
    total_financial_impact = impact_result.scalar() or 0

    return {
        "total": total,
        "by_status": by_status,
        "by_type": by_type,
        "total_financial_impact": float(total_financial_impact),
    }
